vllm/v1/structured_output/bitmasking_grammar.py

```diff
@@ -0,0 +1,383 @@
+                structured_output_request_ids.get(req_id)
+            if mask_index is None:
+                continue
+            batch_index = input_batch.req_id_to_index[req_id]
+            struct_out_indices.append(batch_index)
+            mask_indices.append(mask_index)
+        self.grammar_bitmask_cpu[struct_out_indices] = torch.from_numpy(
+            grammar_bitmask[mask_indices])
+        # It's not guaranteed that all requests in this batch require
+        # structured output, so create a bool tensor to represent
+        # the requests that need structured output.
+        struct_out_indices = torch.tensor(struct_out_indices, dtype=torch.long)
+        self.require_structured_out_cpu[struct_out_indices] = True
+        return self.require_structured_out_cpu[:num_reqs].to(logits.device), \
+            self.grammar_bitmask_cpu[:num_reqs].to(logits.device), \
+            self.structured_decode_arange.to(logits.device)
+
+    @torch.compile(backend="openxla", fullgraph=True, dynamic=False)
+    def structured_decode_tpu(self, require_struct_decoding: torch.Tensor,
+                              grammar_bitmask: torch.Tensor,
+                              logits: torch.Tensor, arange: torch.Tensor):
+        """Applies structured decoding by modifying logits in-place 
+        where required.
+        
+        Args:
+            require_struct_decoding: [B] boolean tensor indicating 
+                which batch items need structured decoding
+            grammar_bitmask: [B, vocab_size//32] packed bit tensor 
+                containing valid token masks
+            logits: [B, vocab_size] tensor to modify in-place
+            arange: [32] tensor for bit unpacking, contains values [0..31]
+        """
+        assert (logits.shape[0] == grammar_bitmask.shape[0])
+
+        # Unpack bits for all batch items at once
+        unpacked_bitmask = (
+            torch.bitwise_right_shift(
+                grammar_bitmask[:, :, None],  # [B, vocab_size//32, 1]
+                arange[None, None, :]  # [1, 1, 32]
+            ) & 1) == 0  # Result: [B, vocab_size//32, 32]
+
+        unpacked_bitmask = unpacked_bitmask.reshape(
+            logits.shape[0], -1)[:, :self.tpu_vocab_size]  # [B, vocab_size]
+
+        # Only apply mask where require_struct_decoding is True
+        mask_to_apply = unpacked_bitmask & \
+            require_struct_decoding[:,None]  # [B, vocab_size]
+
+        # Apply mask in-place
+        logits.masked_fill_(mask_to_apply, -float("inf"))
+
+    def precompile(self, num_reqs_paddings: list[int], device: torch.device,
+                   hidden_states_dtype: torch.dtype):
+        if current_platform.is_tpu():
+            for num_reqs in num_reqs_paddings:
+                dummy_logits = torch.zeros((num_reqs, self.tpu_vocab_size),
+                                           device=device,
+                                           dtype=hidden_states_dtype)
```

Comment:
same thing here, I think it's best to keep the TPU code inside the runner, and then have the runner use the bitmask object to compile. 
I don't think we should scatter the tpu init logic through the code base.